### DISCUSSION TITLE
Signing from native app using the next-auth Rest API

### DISCUSSION THREAD

**Discussion by leo-petrucci** [nextauthjs/next-auth#2621] [Help] [answered]
I'm trying to use my next-auth authentication to sign in from a native app but I'm encountering some problems.

The app is built with Ionic and Capacitor which allow to set the app's hostname to the same hostname as my API, so I don't think there should be any problems with subdomains or cross-domains requests, however I'm having trouble making any requests at all!

From reading the [Rest API docs page](https://next-auth.js.org/getting-started/rest-api), I assumed I would be able to make requests to these endpoints without having to use next-auth's client. But while testing it out with postman I tried to make a post request to `http://localhost:8000/api/auth/signin/twitter`, which just ended up returning me the HTML for the default login page:

![image](https://user-images.githubusercontent.com/23533178/131229036-902c0eb8-7535-405c-8629-e56fd450fa47.png)

Despite getting the csrf token and passing it as form data ([which I think is what I'm supposed to do?](https://next-auth.js.org/getting-started/rest-api#get-apiauthcsrf)) I don't seem to be getting a usable response.

I've also tried it on a live app:

```
  const loginTest = async () => {
    const { csrfToken }: { csrfToken: string } = await fetch(
      "/api/auth/csrf"
    ).then((res) => res.json());

    let formData = new FormData();
    formData.append("csrfToken", csrfToken);
    const data = await fetch("/api/auth/signin/twitter", {
      method: "POST",
      body: formData,
    }).then((res) => res.json());
    console.log(data);
  };
```

But this just ends up logging:
```
Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position 0
```

To the console, which I assume is because it's receiving a bunch of HTML back.

Am I misunderstanding how to use this or is there no way to login from the Rest API?

**Comment by balazsorban44** [comment]
I assume you checked, but you are using port `8000`, and usually, the Next.js default is `3000`. Have you made sure that the `NEXTAUTH_URL` reflects that?

**Reply by leo-petrucci** [reply]
Yeah, the app works just fine if I try to login via the usual method. It's just making requests to it that doesn't seem to work.

![image](https://user-images.githubusercontent.com/23533178/131229783-51ff56a5-7379-47ad-afb7-77ba758e50d5.png)

**Reply by leo-petrucci** [reply]
I've just found #1110, should I assume I'm using the Rest API wrong and this isn't possible at all?

**Comment by balazsorban44** [comment]
Maybe the built-in sign-in page source code would give you an idea?

https://github.com/nextauthjs/next-auth/blob/main/src/server/pages/signin.js

**Reply by leo-petrucci** [reply]
I can get it to work if I put everything in a form and submit it like that, but the same doesn't work by making requests to the endpoint themselves.

Is there a trick to it or is it just not possible to do with requests?
I guess I can probably make it work with the form so it doesn't really matter that much, just interested in knowing.

**Reply by balazsorban44** [reply]
I'll have a closer look when I have some time. Our client code doesn't do magical stuff, it talks to the API routes the same way (doing fetch calls) and if those requests are correct, you should get the proper response, I think.

**Reply by balazsorban44** [reply]
Here is the frontend signin method as well https://github.com/nextauthjs/next-auth/blob/main/src/client/index.js#L177-L238

**Comment by balazsorban44** [comment, marked as answer]
@creativiii I just checked it on Postman, and it worked for me. looks like what is missing for you is `json: true` in your POST body!

![](https://i.imgur.com/JCu8xcY.png)

The URL you are receiving is where you will have to redirect the user to log into their account.

Also, remember to use `x-www-form-urlencoded`

In code, the request should look like this:

```js
const loginTest = async () => {
  const csrfResponse = await fetch("/api/auth/csrf")
  const { csrfToken } = await csrfResponse.json()
  const signInResposne = await fetch("/api/auth/signin/twitter", {
    method: "POST",
    body: new URLSearchParams({ 
      csrfToken,
      json: true
    })
  })
  const { url } = await signInResposne.json()
  console.log(url)
}
```

**Reply by leo-petrucci** [reply]
That explains it! I was just using `formData` and also wasn't passing `json: true`. Thank you for helping, this solved it 😊

**Reply by balazsorban44** [reply]
can you verify that the `csrfToken` is not `null`?

**Reply by leo-petrucci** [reply]
csrfToken was not null, but the problem was caused by not having `json: true` 😀

**Comment by maksymarkhipov** [comment]
@balazsorban44's solution is a good trick, thanks, man. If someone wants to use the solution in the NextJS project (for example, to make implicit auth flow possible), just copy cookies from a csrfToken request (the `csrfResponse` from the solution) into your second request (the `signInResposne` from the solution).

As a result, you will add:
`const cookies = csrfResponse.headers.get('set-cookie')` after a first fetch function
and 
`headers: { 'Cookie': cookies }` in a second fetch function params
